includes/class-reader-activation.php

```diff
@@ -0,0 +1,328 @@
+<?php
+/**
+ * Reader Activation.
+ *
+ * @package Newspack
+ */
+
+namespace Newspack;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Reader Activation Class.
+ */
+final class Reader_Activation {
+
+	const AUTH_INTENTION_COOKIE = 'np_auth_intention';
+
+	/**
+	 * Reader user meta keys.
+	 */
+	const READER         = 'np_reader';
+	const EMAIL_VERIFIED = 'np_reader_email_verified';
+
+	/**
+	 * Initialize hooks.
+	 */
+	public static function init() {
+		if ( self::is_enabled() ) {
+			\add_action( 'clear_auth_cookie', [ __CLASS__, 'clear_auth_intention_cookie' ] );
+			\add_filter( 'login_form_defaults', [ __CLASS__, 'add_auth_intention_to_login_form' ], 20 );
+			\add_action( 'resetpass_form', [ __CLASS__, 'verify_reader_email' ] );
+		}
+	}
+
+	/**
+	 * Whether reader activation is enabled.
+	 *
+	 * @return bool True if reader activation is enabled.
+	 */
+	public static function is_enabled() {
+		$is_enabled = defined( 'NEWSPACK_EXPERIMENTAL_READER_ACTIVATION' ) && NEWSPACK_EXPERIMENTAL_READER_ACTIVATION;
+
+		/**
+		 * Filters whether reader activation is enabled.
+		 *
+		 * @param bool $is_enabled Whether reader activation is enabled.
+		 */
+		return \apply_filters( 'newspack_reader_activation_enabled', $is_enabled );
+	}
+
+	/**
+	 * Add auth intention email to login form defaults.
+	 *
+	 * @param array $defaults Login form defaults.
+	 *
+	 * @return array
+	 */
+	public static function add_auth_intention_to_login_form( $defaults ) {
+		$email = self::get_auth_intention();
+		if ( ! empty( $email ) ) {
+			$defaults['label_username'] = __( 'Email address', 'newspack' );
+			$defaults['value_username'] = $email;
```

Comment:
The hook is only used by `wp_login_form()`, the `wp-login.php` page does not have such functionality. WooCommerce too, doesn't have any hook to add defaults to the form. The only way would be hardcoding it.

Reply:
There's no hook, but there's a way (rough sketch):

```php
add_action( 'woocommerce_login_form_start', function(){
	$_POST['username']=$_COOKIE[ Reader_Activation::AUTH_INTENTION_COOKIE ];
} );
add_action( 'woocommerce_login_form_end', function(){
	$_POST['username']='';
} );
```

This can be tackled later though. [Something similar](https://github.com/WordPress/WordPress/blob/4fc2b46aa0f58d03b4b10c5109e2c354bc9aebd7/wp-login.php#L824-L833) for `/wp-login.php` is not doable though, because after messing with `$_POST` there's a login attempt, which fails since there's no password.

Reply:
I agree we can look at it later. I'm not a fan of populating `$_POST` values outside of a `POST` request.

We are likely to use our own custom-made flow for authentication and rarely fall into `wp-login.php` or WC's login.